framework/audio/internal/sequencer.cpp

```diff
@@ -0,0 +1,262 @@
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "sequencer.h"
+#include "internal/midiplayer.h"
+#include "internal/audioplayer.h"
+
+using namespace mu::audio;
+
+Sequencer::Sequencer()
+{
+    m_clock = std::make_shared<Clock>();
+    m_clock->addSyncCallback([this]() {
+        timeUpdate();
+    });
+    buildRpcReflection();
+}
+
+ISequencer::Status Sequencer::status() const
+{
+    return m_status;
+}
+
+mu::async::Channel<ISequencer::Status> Sequencer::statusChanged() const
+{
+    return m_statusChanged;
+}
+
+bool Sequencer::play()
+{
+    setStatus(PLAYING);
+    for (auto& track : m_tracks) {
+        track.second->play();
+    }
+    m_clock->start();
+    return true;
+}
+
+void Sequencer::pause()
+{
+    setStatus(STOPED);
+    for (auto& track : m_tracks) {
+        track.second->stop();
+    }
+    m_clock->stop();
+}
+
+void Sequencer::stop()
+{
+    setStatus(STOPED);
+    m_clock->stop();
+    for (auto& track : m_tracks) {
+        track.second->stop();
+    }
+    rewind();
+}
+
+void Sequencer::initMIDITrack(ISequencer::track_id id)
+{
+    createMIDITrack(id);
+}
+
+void Sequencer::setMIDITrack(ISequencer::track_id id, const std::shared_ptr<mu::midi::MidiStream>& stream)
+{
+    midi_track_t track = midiTrack(id);
+    if (!track) {
+        track = createMIDITrack(id);
+    }
+    track->loadMIDI(stream);
+}
+
+void Sequencer::setAudioTrack(ISequencer::track_id id, const std::shared_ptr<audio::IAudioStream>& stream)
+{
+    audio_track_t track = audioTrack(id);
+    if (!track) {
+        track = createAudioTrack(id);
+    }
+    track->load(stream);
+}
+
+mu::async::Channel<ISequencer::audio_track_t> Sequencer::audioTrackAdded() const
+{
+    return m_audioTrackAdded;
+}
+
+void Sequencer::initAudioTrack(ISequencer::track_id id)
+{
+    createAudioTrack(id);
+}
+
+mu::async::Channel<mu::midi::tick_t> Sequencer::midiTickPlayed(ISequencer::track_id id) const
+{
+    if (midi_track_t track = midiTrack(id)) {
+        return track->tickPlayed();
+    }
+    return {};
+}
+
+mu::async::Notification Sequencer::positionChanged() const
+{
+    return m_positionChanged;
+}
+
+float Sequencer::playbackPosition() const
+{
+    return clock()->timeInSeconds();
+}
+
+void Sequencer::seek(unsigned long miliseconds)
+{
+    m_clock->seekMiliseconds(miliseconds);
+    for (auto& track : m_tracks) {
+        track.second->seek(miliseconds);
+    }
+    m_positionChanged.notify();
+}
+
+void Sequencer::rewind()
+{
+    seek(0);
+}
+
+void Sequencer::setLoop(unsigned int from, unsigned int to)
+{
+    m_loopStart = from;
+    m_loopEnd = to;
+}
+
+void Sequencer::unsetLoop()
+{
+    m_loopStart.reset();
+    m_loopEnd.reset();
+}
+
+std::shared_ptr<Clock> Sequencer::clock() const
+{
+    return m_clock;
+}
+
+void Sequencer::timeUpdate()
+{
+    if (m_loopStart.has_value() && m_loopEnd.has_value()) {
```

Comment:
This code is not thread-safe.
This method is executed in the audio worker thread, but the loop variables can be changed in the main thread (judging by the use of the sequencer in PlaybackController, direct calls to the methods of this class from the main thread are performed)

Reply:
fixed https://github.com/musescore/MuseScore/pull/7266